### PR TITLE
Rework launchers

### PR DESCRIPTION
**Pull request by pchote**
This is the third and final PR backporting the project structure changes from #17989 for the playtest.
The first (new) commit splits the OpenRA.exe entry point into its own project, and changes OpenRA.Game to a library.
The second commit overhauls the windows launchers, taking a new approach that i'm surprised nobody considered earlier, using assembly attributes that can be passed to msbuild at compile time to avoid editing source files.

~Depends on #18846.~
Supersedes #18582.

Test builds: https://github.com/pchote/OpenRA/releases/tag/devtest-20201122

### CODE REVIEW

**Comment by pchote** [comment]
Rebased.

File: packaging/windows/buildpackage.sh [hunk 1]
```diff
@@ -60,13 +57,12 @@ function build_platform()
 	PLATFORM="${1}"
 
 	echo "Building core files (${PLATFORM})"
+	TARGETPLATFORM="TARGETPLATFORM=win-${PLATFORM}"
```
Comment (teinarss) [inline packaging/windows/buildpackage.sh]: You could make the same changes i did in https://github.com/teinarss/OpenRA/blob/netstandard21/packaging/windows/buildpackage.sh to simplify it more.

Comment (pchote) [reply packaging/windows/buildpackage.sh]: I actually did originally have this, but dropped it when I was having problems getting the build to work. That ended up being something unrelated (and really silly), so I have added it back 👍

**Comment by pchote** [comment]
Updated.

**Review by abcdefg30 — DISMISSED** [review]
Lgtm and works, but the packaging changes are untested. One thing I did notice is that the error dialogue doesn't always open in the foreground. Is there some sort of setting to do that?

**Comment by pchote** [comment]
The packaging changes (except for the latest revision, but I tested that locally and verified that both the 64 and 32 bit installers work) are tested by the devtest tag, with travis logs at https://travis-ci.com/github/pchote/OpenRA/builds/203989439

I could not find any way to force the error dialog on top. I don't think this is supported in SDL 😞

**Comment by abcdefg30** [comment]
https://stackoverflow.com/questions/11512373/findwindow-and-setforegroundwindow-alternatives/12758966#12758966 seems to offer a solution. The crash dialogue seems to retain the name from the exe, so I could get one to the front locally by using "TiberianDawn" as process name.

**Comment by pchote** [comment]
That won't work because `SDL.SDL_ShowMessageBox` blocks until the dialog is closed, so we don't have the chance to run any code while the dialog is open. Calling `SetForegroundWindow` before opening the message box doesn't work either because the launcher's `MainWindowHandle` is 0 (i.e. null) because it doesn't have any windows.

**Comment by pchote** [comment]
I tried calling `SetForegroundWindow` on a different thread in several different ways, but couldn't get it to work (it would blink the icon in the taskbar, so *something* was happening, but the dialog was not brought to the front). I suspect this may be because the message box is not the `MainWindowHandle`, but I don't know how to get any other window handles.

I'm out of ideas, so can't proceed any further with that suggestion/request without help.

If anyone wants to try modifying the code to test ideas you can compile the TD launcher with
```
msbuild -t:Build "OpenRA.WindowsLauncher/OpenRA.WindowsLauncher.csproj" -restore -p:Configuration=Release -p:TargetPlatform="win-x64" -p:LauncherName="TiberianDawn" -p:ModID="cnc" -p:DisplayName="Tiberian Dawn" -p:FaqUrl="http://wiki.openra.net/FAQ"
```

and then copy `TiberianDawn.exe` from the bin dir into an existing Windows install for testing.

**Comment by pchote** [comment]
The behaviour I mentioned above is explained by [this link](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setforegroundwindow):
> An application cannot force a window to the foreground while the user is working with another window. Instead, Windows flashes the taskbar button of the window to notify the user. 

So it sounds like we should be expecting this to fail.

**Comment by pchote** [comment]
The trick is that the game process needs to grant permission for the launcher to set the foreground window before it crashes. Added a new commit to fix this, which you can test using the msbuild command above.

**Comment by abcdefg30** [comment]
That did the trick, good job. :+1:

**Comment by pchote** [comment]
Open question: now that the platform launchers link against `OpenRA.Game.dll`, should we stop shipping the generic `OpenRA.exe` launcher in the windows builds? This will stop people from "crashing" if they try to run OpenRA.exe through the GUI.

**Comment by abcdefg30** [comment]
If there is no need to ship the generic exe, then we should drop it.

**Comment by pchote** [comment]
Ok, done. I also scope creeped in replacing MakeLAA.cs/exe with a python script to remove another change from #17989.

New packages will be available from https://github.com/pchote/OpenRA/releases/tag/devtest-20201127-2

File: OpenRA.WindowsLauncher/OpenRA.WindowsLauncher.csproj [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    <OutputType>winexe</OutputType>
+    <TargetFramework>net472</TargetFramework>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+    <Optimize>true</Optimize>
+    <UseVSHostingProcess>false</UseVSHostingProcess>
+    <LangVersion>7.3</LangVersion>
+    <DebugSymbols>true</DebugSymbols>
+    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+    <AutoGenerateBindingRedirects>false</AutoGenerateBindingRedirects>
+    <RootNamespace>OpenRA</RootNamespace>
```
Comment (teinarss) [inline OpenRA.WindowsLauncher/OpenRA.WindowsLauncher.csproj]: Remove

File: OpenRA.Launcher/OpenRA.Launcher.csproj [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    <OutputType>Exe</OutputType>
+    <TargetFramework>net472</TargetFramework>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+    <Optimize>true</Optimize>
+    <UseVSHostingProcess>false</UseVSHostingProcess>
+    <LangVersion>7.3</LangVersion>
+    <DebugSymbols>true</DebugSymbols>
+    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+    <AutoGenerateBindingRedirects>false</AutoGenerateBindingRedirects>
+    <RootNamespace>OpenRA</RootNamespace>
```
Comment (teinarss) [inline OpenRA.Launcher/OpenRA.Launcher.csproj]: Remove

**Comment by pchote** [comment]
Updated. I also removed the `Release-x86` configurations because I remembered that #17744 removed them everywhere else.

**Review by teinarss — APPROVED** [review]

**Comment by teinarss** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/6eab03c56ab1600897c0d2dc3d90254e62cae51b...27e356d24da4d6b056e6031a32237dc2488a8ff6

**Pull request merged**